File: Parallel-Java/src/parallel/linq101/SetOperations.java
package parallel.linq101;

import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class SetOperations {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public int distinctSyntax() {
        int[] factorsOf300 = {2, 2, 3, 5, 5};

        var uniqueFactors = Arrays.stream(factorsOf300).distinct().toArray();

        System.out.println("Prime factors of 300:");
        for (var f : uniqueFactors)
            System.out.println(f);

        return 0;
    }

    public int distinctPropertyValues() {
        List<Product> products = getProductList();

        var categoryNames = products.stream().map(p -> p.getCategory())
                .distinct().collect(Collectors.toList());

        System.out.println("Category names:");
        for (var n : categoryNames)
            System.out.println(n);
        return 0;
    }

    public int unionSyntax() {
        int[] numbersA = {0, 2, 4, 5, 6, 8, 9};
        int[] numbersB = {1, 3, 5, 7, 8};

        var uniqueNumbers = IntStream.concat(Arrays.stream(numbersA),
                Arrays.stream(numbersB)).distinct().toArray();

        System.out.println("Unique numbers from both arrays:");
        for (var n : uniqueNumbers)
            System.out.println(n);

        return 0;
    }

    public int unionOfQueryResults() {
        List<Product> products = getProductList();
        List<Customer> customers = getCustomerList();

        var productFirstChars = products.stream().map(p -> p.getProductName().charAt(0));
        var customerFirstChars = customers.stream().map(c -> c.getCompanyName().charAt(0));
        var uniqueFirstChars = Stream.concat(productFirstChars, customerFirstChars)
                .distinct().collect(Collectors.toList());

        System.out.println("Unique first letters from Product names and Customer names:");
        for (var ch : uniqueFirstChars)
            System.out.println(ch);

        return 0;
    }

    public int intersectSynxtax() {
        int[] numbersA = {0, 2, 4, 5, 6, 8, 9};
        int[] numbersB = {1, 3, 5, 7, 8};

        var commonNumbers = Arrays.stream(numbersA).filter(a ->
                Arrays.stream(numbersB).anyMatch(b->b==a)).toArray();

        System.out.println("Common numbers shared by both arrays:");
        for (var n : commonNumbers)
            System.out.println(n);
        return 0;
    }

    public int intersectQueryResults() {
        List<Product> products = getProductList();
        List<Customer> customers = getCustomerList();

        var productFirstChars = products.stream().map(p -> p.getProductName().charAt(0));
        var customerFirstChars = customers.stream().map(c -> c.getCompanyName().charAt(0));
        var commonFirstChars = productFirstChars.filter(p -> customerFirstChars.anyMatch(c -> c == p))
                .collect(Collectors.toList());

        System.out.println("Common first letters from Product names and Customer names:");
        for (var ch : commonFirstChars)
            System.out.println(ch);
        return 0;
    }

    public int differenceOfSets() {
        int[] numbersA = {0, 2, 4, 5, 6, 8, 9};
        int[] numbersB = {1, 3, 5, 7, 8};

        var aOnlyNumbers = Arrays.stream(numbersA).filter(a ->
                Arrays.stream(numbersB).allMatch(b->b!=a)).toArray();

        System.out.println("Numbers in first array but not second array:");
        for(var n : aOnlyNumbers)
            System.out.println(n);

        return 0;
    }

    public int DifferenceOfQueries() {

        List<Product> products = getProductList();
        List<Customer> customers = getCustomerList();

        var productFirstChars = products.stream().map(p -> p.getProductName().charAt(0));
        var customerFirstChars = customers.stream().map(c -> c.getCompanyName().charAt(0));
        var productOnlyFirstChars = productFirstChars.filter(p -> customerFirstChars.allMatch(c -> c != p))
                .collect(Collectors.toList());

        System.out.println("First letters from Product names, but not from Customer names:");
        for(var n : productOnlyFirstChars)
            System.out.println(n);

        return 0;
    }
}

File: Parallel-Csharp/src/Parallel/Linq101/SetOperations.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101;

public class SetOperations
{
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    public int DistinctSyntax()
    {
        int[] factorsOf300 = { 2, 2, 3, 5, 5 };

        var uniqueFactors = factorsOf300.Distinct().ToArray();

        Console.WriteLine("Prime factors of 300:");
        foreach (var f in uniqueFactors)
            Console.WriteLine(f);

        return 0;
    }

    public int DistinctPropertyValues()
    {
        List<Product> products = GetProductList();

        var categoryNames = products.Select(p => p.Category)
            .Distinct().ToList();

        Console.WriteLine("Category names:");
        foreach (var n in categoryNames)
            Console.WriteLine(n);
        return 0;
    }

    public int UnionSyntax()
    {
        int[] numbersA = { 0, 2, 4, 5, 6, 8, 9 };
        int[] numbersB = { 1, 3, 5, 7, 8 };

        var uniqueNumbers = numbersA.Union(numbersB).ToArray();

        Console.WriteLine("Unique numbers from both arrays:");
        foreach (var n in uniqueNumbers)
            Console.WriteLine(n);

        return 0;
    }

    public int UnionOfQueryResults()
    {
        List<Product> products = GetProductList();
        List<Customer> customers = GetCustomerList();

        var productFirstChars = products.Select(p => p.ProductName[0]);
        var customerFirstChars = customers.Select(c => c.CompanyName[0]);
        var uniqueFirstChars = productFirstChars.Union(customerFirstChars).ToList();

        Console.WriteLine("Unique first letters from Product names and Customer names:");
        foreach (var ch in uniqueFirstChars)
            Console.WriteLine(ch);

        return 0;
    }

    public int IntersectSynxtax()
    {
        int[] numbersA = { 0, 2, 4, 5, 6, 8, 9 };
        int[] numbersB = { 1, 3, 5, 7, 8 };

        var commonNumbers = numbersA.Intersect(numbersB).ToArray();

        Console.WriteLine("Common numbers shared by both arrays:");
        foreach (var n in commonNumbers)
            Console.WriteLine(n);
        return 0;
    }

    public int IntersectQueryResults()
    {
        List<Product> products = GetProductList();
        List<Customer> customers = GetCustomerList();

        var productFirstChars = products.Select(p => p.ProductName[0]);
        var customerFirstChars = customers.Select(c => c.CompanyName[0]);
        var commonFirstChars = productFirstChars.Intersect(customerFirstChars).ToList();

        Console.WriteLine("Common first letters from Product names and Customer names:");
        foreach (var ch in commonFirstChars)
            Console.WriteLine(ch);
        return 0;
    }

    public int DifferenceOfSets()
    {
        int[] numbersA = { 0, 2, 4, 5, 6, 8, 9 };
        int[] numbersB = { 1, 3, 5, 7, 8 };

        var aOnlyNumbers = numbersA.Except(numbersB).ToArray();

        Console.WriteLine("Numbers in first array but not second array:");
        foreach (var n in aOnlyNumbers)
            Console.WriteLine(n);

        return 0;
    }

    public int DifferenceOfQueries()
    {

        List<Product> products = GetProductList();
        List<Customer> customers = GetCustomerList();
        
        var productFirstChars = products.Select(p => p.ProductName[0]);
        var customerFirstChars = customers.Select(c => c.CompanyName[0]);
        var productOnlyFirstChars = productFirstChars.Except(customerFirstChars).ToList();

        Console.WriteLine("First letters from Product names, but not from Customer names:");
        foreach (var ch in productOnlyFirstChars)
            Console.WriteLine(ch);

        return 0;
    }
}